In [1]:
import pandas as pd

#import all files
df1 = pd.read_csv("North_America_Airbus.csv") 
df2 = pd.read_csv("North_America_Boeing_CFM.csv")
df3 = pd.read_csv("North_America_Boeing_Rest.csv")
df4 = pd.read_csv("EasternEU_Airbus.csv")
df5 = pd.read_csv("EasternEU_Boeing.csv")
df6 = pd.read_csv("West_EU_Airbus.csv")
df7 = pd.read_csv("West_EU_Boeing.csv")
df8 = pd.read_csv("Latin_Airbus.csv")
df9 = pd.read_csv("Latin_Boeing.csv")
iso = pd.read_csv("paises.csv")

#combine all dataframes into one
df_aw = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9], ignore_index=True)
df_iso = iso



In [2]:
df_aw.head()

Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Aircraft Variant Detail Engine Manufacturer Engine Type  \
0                     212                 CFM       CFM56   
1                     214                 CFM       CFM56   
2                     112                 CFM       CFM56   
3                     214                 CFM       CFM56   
4                     214                 CFM       CFM56   

  Engine Model Variant Engine Variant Detail APU Full Name  \
0                   5A                   5A3             -   
1                   5B                 5B4/P             -   
2                   5B                 5B6/P             -   
3                   5B                 5B4/P             -   
4                   5B                 5B4/P             -   

  Aircraft Registration Number  ...                   Operator Phone  \
0                       N364NW  ...                     404-715-2600   
1                       C-FGJI  ...  888-247-2262 (Customer Service)   
2                       N302NV  ...                     702-851-7300   
3                       N297NV  ...                     702-851-7300   
4                       N283NV  ...                     702-851-7300   

                       Operator Email               Registered Owner/Manager  \
0                                   -                        DELTA AIR LINES   
1                                   -                AERGEN AVIATION FINANCE   
2  customerrelations@allegiantair.com  SUNRISE ASSET MANAGEMENT LLC [NV-USA]   
3  customerrelations@allegiantair.com                 BANK OF UTAH [TRUSTEE]   
4  customerrelations@allegiantair.com  SUNRISE ASSET MANAGEMENT LLC [NV-USA]   

  In Service On Order Stored  Hours Cycles Age (Since Delivery)  \
0        1.0      NaN    NaN  75833  35997                 26.8   
1        1.0      NaN    NaN  70865  27058                 23.6   
2        1.0      NaN    NaN  51115  31015                 20.8   
3        1.0      NaN    NaN  55338  18967                 19.6   
4        1.0      NaN    NaN  53381  21646                 19.2   

  Number of Engines for Operators  
0                               2  
1                               2  
2                               2  
3                               2  
4                               2  

[5 rows x 27 columns]

In [3]:
df_iso.head()

nombre         name          nom  iso2  iso3  phone_code
0  Afganistán  Afghanistan  Afghanistan    AF   AFG          93
1     Albania      Albania      Albanie    AL   ALB         355
2    Alemania      Germany    Allemagne    DE   DEU          49
3     Andorra      Andorra      Andorra    AD   AND         376
4      Angola       Angola       Angola    AO   AGO         244

In [4]:
#Clean column names & database entries: PAISES.CSV
df_iso.columns = df_iso.columns.str.strip()

df_iso['name'] = df_iso['name'].astype('string').str.strip().str.lower()

df_iso.head()


nombre         name          nom iso2 iso3 phone_code
0  Afganistán  afghanistan  Afghanistan   AF  AFG         93
1     Albania      albania      Albanie   AL  ALB        355
2    Alemania      germany    Allemagne   DE  DEU         49
3     Andorra      andorra      Andorra   AD  AND        376
4      Angola       angola       Angola   AO  AGO        244

In [5]:
#Review column required within the exported files
df_iso = df_iso[['name','iso3']]
df_iso.head()

name iso3
0  afghanistan  AFG
1      albania  ALB
2      germany  DEU
3      andorra  AND
4       angola  AGO

In [6]:
#Review column required within the exported files
df_aw = df_aw[['Aircraft Manufacturer', 'Aircraft Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]

df_aw.head()

Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Engine Manufacturer Engine Type Engine Model Variant  \
0                 CFM       CFM56                   5A   
1                 CFM       CFM56                   5B   
2                 CFM       CFM56                   5B   
3                 CFM       CFM56                   5B   
4                 CFM       CFM56                   5B   

  Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                       N364NW      In Service             03/04/1999   
1                       C-FGJI      In Service             06/06/2002   
2                       N302NV      In Service             02/07/2005   
3                       N297NV      In Service             05/30/2006   
4                       N283NV      In Service             10/30/2006   

                Operator Operator Country  Hours  Cycles  \
0  Delta Air Lines, Inc.    United States  75833   35997   
1             Air Canada           Canada  70865   27058   
2          Allegiant Air    United States  51115   31015   
3          Allegiant Air    United States  55338   18967   
4          Allegiant Air    United States  53381   21646   

   Age (Since Delivery)  Number of Engines for Operators  
0                  26.8                                2  
1                  23.6                                2  
2                  20.8                                2  
3                  19.6                                2  
4                  19.2                                2

In [7]:
#filter the aircraft status to In Service
df_aw = df_aw[df_aw['Aircraft Status'] == 'In Service']

df_aw.head()

Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Engine Manufacturer Engine Type Engine Model Variant  \
0                 CFM       CFM56                   5A   
1                 CFM       CFM56                   5B   
2                 CFM       CFM56                   5B   
3                 CFM       CFM56                   5B   
4                 CFM       CFM56                   5B   

  Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                       N364NW      In Service             03/04/1999   
1                       C-FGJI      In Service             06/06/2002   
2                       N302NV      In Service             02/07/2005   
3                       N297NV      In Service             05/30/2006   
4                       N283NV      In Service             10/30/2006   

                Operator Operator Country  Hours  Cycles  \
0  Delta Air Lines, Inc.    United States  75833   35997   
1             Air Canada           Canada  70865   27058   
2          Allegiant Air    United States  51115   31015   
3          Allegiant Air    United States  55338   18967   
4          Allegiant Air    United States  53381   21646   

   Age (Since Delivery)  Number of Engines for Operators  
0                  26.8                                2  
1                  23.6                                2  
2                  20.8                                2  
3                  19.6                                2  
4                  19.2                                2

In [8]:
#Extract AC type from Aircraft Type column, where aircraft manufacturer is Airbus left 4 digits in aircraft type and boeing left 3 digits
df_aw['AC Type'] = df_aw.apply(lambda row: row['Aircraft Type'][:4] if 'Airbus' in row['Aircraft Manufacturer'] else row['Aircraft Type'][:3], axis=1)

df_aw.head()


Aircraft Manufacturer Aircraft Type Aircraft Model Variant  \
0                Airbus      A320-200                    200   
1                Airbus      A320-200                    200   
2                Airbus      A319-100                    100   
3                Airbus      A320-200                    200   
4                Airbus      A320-200                    200   

  Engine Manufacturer Engine Type Engine Model Variant  \
0                 CFM       CFM56                   5A   
1                 CFM       CFM56                   5B   
2                 CFM       CFM56                   5B   
3                 CFM       CFM56                   5B   
4                 CFM       CFM56                   5B   

  Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                       N364NW      In Service             03/04/1999   
1                       C-FGJI      In Service             06/06/2002   
2                       N302NV      In Service             02/07/2005   
3                       N297NV      In Service             05/30/2006   
4                       N283NV      In Service             10/30/2006   

                Operator Operator Country  Hours  Cycles  \
0  Delta Air Lines, Inc.    United States  75833   35997   
1             Air Canada           Canada  70865   27058   
2          Allegiant Air    United States  51115   31015   
3          Allegiant Air    United States  55338   18967   
4          Allegiant Air    United States  53381   21646   

   Age (Since Delivery)  Number of Engines for Operators AC Type  
0                  26.8                                2    A320  
1                  23.6                                2    A320  
2                  20.8                                2    A319  
3                  19.6                                2    A320  
4                  19.2                                2    A320

In [9]:
#Review unique values for AC Type
df_aw['AC Type'].unique()

array(['A320', 'A319', 'A321', 'A330', 'A300', 'A220', 'A350', '737',
       '707', '727', '767', '757', '747', '777', 'MD-', '717', '787',
       'A340', 'A380', 'A310'], dtype=object)

In [10]:
#Exclude values "MD-" from column AC type
df_aw = df_aw[~df_aw['AC Type'].str.startswith('MD-', na=False)]
df_aw['AC Type'].unique()

array(['A320', 'A319', 'A321', 'A330', 'A300', 'A220', 'A350', '737',
       '707', '727', '767', '757', '747', '777', '717', '787', 'A340',
       'A380', 'A310'], dtype=object)

In [11]:
#Reorder columns
df_aw = df_aw[['Aircraft Manufacturer', 'AC Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]
df_aw

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
19462                Boeing     737                  Max 8   
19463                Boeing     737                  Max 8   
19464                Boeing     737                  Max 9   
19465                Boeing     737                  Max 8   
19466                Boeing     737                  Max 8   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
19462                 CFM        LEAP                   1B   
19463                 CFM        LEAP                   1B   
19464                 CFM        LEAP                   1B   
19465                 CFM        LEAP                   1B   
19466                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             03/04/1999   
1                           C-FGJI      In Service             06/06/2002   
2                           N302NV      In Service             02/07/2005   
3                           N297NV      In Service             05/30/2006   
4                           N283NV      In Service             10/30/2006   
...                            ...             ...                    ...   
19462                   HP-9809CMP      In Service             09/16/2025   
19463                   HP-9812CMP      In Service             09/25/2025   
19464                       N959AM      In Service             09/30/2025   
19465                   HP-9811CMP      In Service             09/30/2025   
19466                       XA-AMH      In Service             10/15/2025   

                    Operator Operator Country  Hours  Cycles  \
0      Delta Air Lines, Inc.    United States  75833   35997   
1                 Air Canada           Canada  70865   27058   
2              Allegiant Air    United States  51115   31015   
3              Allegiant Air    United States  55338   18967   
4              Allegiant Air    United States  53381   21646   
...                      ...              ...    ...     ...   
19462          Copa Airlines           Panama      0       0   
19463          Copa Airlines           Panama      0       0   
19464             Aeromexico           Mexico      0       0   
19465          Copa Airlines           Panama      0       0   
19466             Aeromexico           Mexico      0       0   

       Age (Since Delivery)  Number of Engines for Operators  
0                      26.8                                2  
1                      23.6                                2  
2                      20.8                                2  
3                      19.6                                2  
4                      19.2                                2  
...                     ...                              ...  
19462                   0.3                                2  
19463                   0.2                                2  
19464                   0.3                                2  
19465                   0.3                                2  
19466                   0.5                                2  

[12973 rows x 15 columns]

In [12]:
#Clean database entries for 'Operator Country':
df_aw['Operator Country'] = df_aw['Operator Country'].astype('string').str.strip().str.lower()

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             03/04/1999  Delta Air Lines, Inc.   
1      In Service             06/06/2002             Air Canada   
2      In Service             02/07/2005          Allegiant Air   
3      In Service             05/30/2006          Allegiant Air   
4      In Service             10/30/2006          Allegiant Air   

  Operator Country  Hours  Cycles  Age (Since Delivery)  \
0    united states  75833   35997                  26.8   
1           canada  70865   27058                  23.6   
2    united states  51115   31015                  20.8   
3    united states  55338   18967                  19.6   
4    united states  53381   21646                  19.2   

   Number of Engines for Operators  
0                                2  
1                                2  
2                                2  
3                                2  
4                                2

In [13]:
#Merge df_aw with iso to get the ISO3 code for each country in 'Operator Country' column
df_aw = df_aw.merge(df_iso, how='left', left_on='Operator Country', right_on='name')

df_aw.head()


Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             03/04/1999  Delta Air Lines, Inc.   
1      In Service             06/06/2002             Air Canada   
2      In Service             02/07/2005          Allegiant Air   
3      In Service             05/30/2006          Allegiant Air   
4      In Service             10/30/2006          Allegiant Air   

  Operator Country  Hours  Cycles  Age (Since Delivery)  \
0    united states  75833   35997                  26.8   
1           canada  70865   27058                  23.6   
2    united states  51115   31015                  20.8   
3    united states  55338   18967                  19.6   
4    united states  53381   21646                  19.2   

   Number of Engines for Operators           name iso3  
0                                2  united states  USA  
1                                2         canada  CAN  
2                                2  united states  USA  
3                                2  united states  USA  
4                                2  united states  USA

In [14]:
#Reorder columns
df_aw = df_aw = df_aw[['Aircraft Manufacturer', 'AC Type', 'Aircraft Model Variant', 'Engine Manufacturer', 'Engine Type','Engine Model Variant',
       'Aircraft Registration Number','Aircraft Status', 'Aircraft Delivery Date','Operator', 'Operator Country', 'iso3','Hours','Cycles', 'Age (Since Delivery)', 'Number of Engines for Operators']]

df_aw

Aircraft Manufacturer AC Type Aircraft Model Variant  \
0                    Airbus    A320                    200   
1                    Airbus    A320                    200   
2                    Airbus    A319                    100   
3                    Airbus    A320                    200   
4                    Airbus    A320                    200   
...                     ...     ...                    ...   
12968                Boeing     737                  Max 8   
12969                Boeing     737                  Max 8   
12970                Boeing     737                  Max 9   
12971                Boeing     737                  Max 8   
12972                Boeing     737                  Max 8   

      Engine Manufacturer Engine Type Engine Model Variant  \
0                     CFM       CFM56                   5A   
1                     CFM       CFM56                   5B   
2                     CFM       CFM56                   5B   
3                     CFM       CFM56                   5B   
4                     CFM       CFM56                   5B   
...                   ...         ...                  ...   
12968                 CFM        LEAP                   1B   
12969                 CFM        LEAP                   1B   
12970                 CFM        LEAP                   1B   
12971                 CFM        LEAP                   1B   
12972                 CFM        LEAP                   1B   

      Aircraft Registration Number Aircraft Status Aircraft Delivery Date  \
0                           N364NW      In Service             03/04/1999   
1                           C-FGJI      In Service             06/06/2002   
2                           N302NV      In Service             02/07/2005   
3                           N297NV      In Service             05/30/2006   
4                           N283NV      In Service             10/30/2006   
...                            ...             ...                    ...   
12968                   HP-9809CMP      In Service             09/16/2025   
12969                   HP-9812CMP      In Service             09/25/2025   
12970                       N959AM      In Service             09/30/2025   
12971                   HP-9811CMP      In Service             09/30/2025   
12972                       XA-AMH      In Service             10/15/2025   

                    Operator Operator Country iso3  Hours  Cycles  \
0      Delta Air Lines, Inc.    united states  USA  75833   35997   
1                 Air Canada           canada  CAN  70865   27058   
2              Allegiant Air    united states  USA  51115   31015   
3              Allegiant Air    united states  USA  55338   18967   
4              Allegiant Air    united states  USA  53381   21646   
...                      ...              ...  ...    ...     ...   
12968          Copa Airlines           panama  PAN      0       0   
12969          Copa Airlines           panama  PAN      0       0   
12970             Aeromexico           mexico  MEX      0       0   
12971          Copa Airlines           panama  PAN      0       0   
12972             Aeromexico           mexico  MEX      0       0   

       Age (Since Delivery)  Number of Engines for Operators  
0                      26.8                                2  
1                      23.6                                2  
2                      20.8                                2  
3                      19.6                                2  
4                      19.2                                2  
...                     ...                              ...  
12968                   0.3                                2  
12969                   0.2                                2  
12970                   0.3                                2  
12971                   0.3                                2  
12972                   0.5                                2  

[1297

In [15]:
#Convert 'Aircraft Delivery Date' to datetime format and save the year
df_aw['Aircraft Delivery Date'] = pd.to_datetime(df_aw['Aircraft Delivery Date'], errors='coerce', dayfirst=True)
df_aw['Delivery Year'] = df_aw['Aircraft Delivery Date'].dt.year.astype('Int64')

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year  
0                                2           1999  
1                                2           2002  
2                                2           2005  
3                                2           <NA>  
4                                2           <NA>

In [16]:
#Filling missing values in 'Delivery Year' with actual year minus age(since delivery)

current_year = pd.Timestamp.now().year

# 1) Asegura tipos numéricos
df_aw['Age (Since Delivery)'] = pd.to_numeric(df_aw['Age (Since Delivery)'], errors='coerce')
df_aw['Delivery Year'] = pd.to_numeric(df_aw['Delivery Year'], errors='coerce')

# 2) Calcula año candidato solo donde falte (puede dar decimales si la edad no es entera)
deliv_year_calc = current_year - df_aw['Age (Since Delivery)']

# 3) Limpia el resultado: rango razonable y redondea a año
deliv_year_calc = deliv_year_calc.where(deliv_year_calc.between(2000, current_year), other=pd.NA)  # descarta valores raros
deliv_year_calc = deliv_year_calc.round()   # o .floor() si prefieres truncar
deliv_year_calc = deliv_year_calc.astype('Int64')  # ya es seguro convertir

# 4) Redondea y convierte también la columna existente (por si vino como float)
deliv_year = df_aw['Delivery Year'].round().astype('Int64')

# 5) Rellena faltantes con el candidato
df_aw['Delivery Year'] = deliv_year.fillna(deliv_year_calc)

df_aw.head()

Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year  
0                                2           1999  
1                                2           2002  
2                                2           2005  
3                                2           2005  
4                                2           2006

In [17]:
!pip install plotly



[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [18]:
import plotly.express as px
import plotly.graph_objects as go

In [19]:
#1. Chart of Aircraft Manufacturer by Aircraft Delivery Year

# Agrupar datos por año y fabricante
evol_manufacturer = (
    df_aw.groupby(['Delivery Year', 'Aircraft Manufacturer'])
         .size()
         .reset_index(name='Number of Aircraft')
         .sort_values('Delivery Year')
)

# Crear el gráfico
fig = px.line(
    evol_manufacturer,
    x='Delivery Year',
    y='Number of Aircraft',
    color='Aircraft Manufacturer',
    markers=True,
    title='Evolución del nº aeronaves en servicio por aircraft manufacturer',
    labels={'Delivery Year': 'Año de Entrega', 'Number of Aircraft': 'Nº de Aeronaves'}
)

fig.update_layout(
    xaxis=dict(dtick=1),  # muestra todos los años si no hay demasiados
    legend_title_text='Fabricante'
)
fig.show()

### TENDENCIAS EN AVIONES EN SERVICIO: BOEING VS AIRBUS

**Eventos socio-económicos**
- **COVID**: Caida de las entregas desde finales 2019-2021
- **Tarifas USA**: Subida de los costes y problemas de SC debido a los aranceles impuestos por Trump en 2025.

**Boeing:**
- **Accidentes mortales**: El 737 MAX sufrió dos accidentes fatales en Indonesia (2018) y Etiopía (2019), causando 346 muertes.
- **Pérdida de confianza**: Durante 2019-2020 se observa una caída significativa en las entregas, visible en el gráfico de evolución.
- **Caida de entregas**: La FAA _(Federal Aeroespace Agency)_ impone limite de producción e inspecciones a Boeing.

**Airbus:**
- **Liderazgo en entregas**: Mantiene una tendencia de crecimiento consistente en entregas anuales.
- **Nuevos contratos importantes**: Capta a 3 grandes clientes de la competencia entre 2022-2024.
- **Ramp up de la produccion**: Crea una linea nueva de producción del A320 equivalente al 737 MAX.


In [20]:
df_aw['Engine Manufacturer'].unique()

#Replace 'Rolls-Royce Deutshland' with 'Rolls-Royce' in 'Engine Manufacturer' column
df_aw['Engine Manufacturer'] = df_aw['Engine Manufacturer'].replace('Rolls-Royce Deutschland', 'Rolls-Royce')


In [21]:
#check unique values again
df_aw['Engine Manufacturer'].unique()

array(['CFM', 'Pratt & Whitney', 'Rolls-Royce', 'GE'], dtype=object)

In [22]:
#Merge columns Aircraft Manufacturer & Engine Manufaturer in a new column 'ManufacturerAC_Engine'

df_aw['ManufacturerAC_Engine'] = df_aw['Aircraft Manufacturer'] + ' - ' + df_aw['Engine Manufacturer']

df_aw.head()


Aircraft Manufacturer AC Type Aircraft Model Variant Engine Manufacturer  \
0                Airbus    A320                    200                 CFM   
1                Airbus    A320                    200                 CFM   
2                Airbus    A319                    100                 CFM   
3                Airbus    A320                    200                 CFM   
4                Airbus    A320                    200                 CFM   

  Engine Type Engine Model Variant Aircraft Registration Number  \
0       CFM56                   5A                       N364NW   
1       CFM56                   5B                       C-FGJI   
2       CFM56                   5B                       N302NV   
3       CFM56                   5B                       N297NV   
4       CFM56                   5B                       N283NV   

  Aircraft Status Aircraft Delivery Date               Operator  \
0      In Service             1999-04-03  Delta Air Lines, Inc.   
1      In Service             2002-06-06             Air Canada   
2      In Service             2005-07-02          Allegiant Air   
3      In Service                    NaT          Allegiant Air   
4      In Service                    NaT          Allegiant Air   

  Operator Country iso3  Hours  Cycles  Age (Since Delivery)  \
0    united states  USA  75833   35997                  26.8   
1           canada  CAN  70865   27058                  23.6   
2    united states  USA  51115   31015                  20.8   
3    united states  USA  55338   18967                  19.6   
4    united states  USA  53381   21646                  19.2   

   Number of Engines for Operators  Delivery Year ManufacturerAC_Engine  
0                                2           1999          Airbus - CFM  
1                                2           2002          Airbus - CFM  
2                                2           2005          Airbus - CFM  
3                                2           2005          Airbus - CFM  
4                                2           2006          Airbus - CFM

In [24]:
#Group by Engine Manufaturer by Operator Country

#1) Agregación

# a) Recuento DISTINCT por matrícula 
agg = (df_aw .dropna(subset=['ManufacturerAC_Engine', 'iso3', 'Operator Country']) 
       .groupby(['iso3','Operator Country', 'ManufacturerAC_Engine'])['Aircraft Registration Number'] 
       .nunique() 
       .reset_index(name='aircraft_count'))

# Filtro top 10 paises:
top_countries = (agg.groupby('iso3')['aircraft_count'].sum() .nlargest(10).index)

agg = agg[agg['iso3'].isin(top_countries)]


#b) Reagrega por motor
agg_engine = (
    df_aw.dropna(subset=['iso3','Operator Country','Engine Manufacturer','Aircraft Registration Number'])
         .groupby(['iso3','Operator Country','Engine Manufacturer'])['Aircraft Registration Number']
         .nunique().reset_index(name='aircraft_count')
)

# Filtra motores principales
top_engines = (
    agg_engine.groupby('Engine Manufacturer')['aircraft_count'].sum()
              .nlargest(6).index)

agg_engine = agg_engine[agg_engine['Engine Manufacturer'].isin(top_engines)]

#2) Filtrar top K combinaciones ManufacturerAC_Engine

TOP_K = 8

top_combo = (
    agg.groupby('ManufacturerAC_Engine')['aircraft_count']
       .sum()
       .nlargest(TOP_K)
       .index
)

agg_top = agg[agg['ManufacturerAC_Engine'].isin(top_combo)].copy()


#3) Mapa

fig = px.scatter_geo(
    agg_top,
    locations='iso3',
    size='aircraft_count',
    color='ManufacturerAC_Engine',
    facet_col='ManufacturerAC_Engine',
    facet_col_wrap=4,         # cuántos por fila
    size_max=24,
    hover_name='Operator Country',
    projection='natural earth',
    title=f'Distribución por país de las top combinaciones de Manufacturer aircraft & Engine'
)
# Limpiar títulos de faceta
fig.for_each_annotation(lambda a: a.update(text=a.text.split('=')[-1]))
fig.update_layout(showlegend=False, margin=dict(l=10, r=10, t=60, b=10))
fig.show()
